### PR TITLE
Sp reimplement

### PR DESCRIPTION
**Pull request by jcapriot**
#### Summary
Reimplements the SP method in SimPEG.

#### PR Checklist
* [x] If this is a work in progress PR, set as a Draft PR
* [x] Linted my code according to the [style guides](https://docs.simpeg.xyz/content/basic/practices.html#style).
* [x] Added [tests](https://docs.simpeg.xyz/content/basic/practices.html#testing) to verify changes to the code.
* [x] Added necessary documentation to any new functions/classes following the
      expect [style](https://docs.simpeg.xyz/content/basic/practices.html#documentation).
* [x] Added relevant method tags (i.e. `GRAV`, `EM`, etc.)
* [x] Marked as ready for review (ff this is was a draft PR), and converted 
      to a Pull Request
* [x] Tagged ``@simpeg/simpeg-developers`` when ready for review.

#### What does this implement/fix?
This reimplements the SP code in SimPEG, making heavy use of the existing DC formulation. It is written that the `q` property is the model.

It also defines a few new maps that are specific to the SP methodology, one relating the secondary current density vector defined at cell-centers to `q` by taking the negative divergence of it. The second relates the hydraulic head to the `q`. These two maps are simple extensions of the `LinearMap`. Together with the simple formulation relating to `q` these should mimic all of the previous functionality of the SP code.

#### Additional information

I went back and forth deciding on the name of the module this lives, i.e. it is either `spontaneous_potential` or `self_potential`. I went in favor of `spontaneous` purely because the word `self` has a common meaning in python programming. But I'm happy either way.

### CODE REVIEW

**Comment by jcapriot** [comment]
@sgkang Here is my implementation of the SP simulation. Very lightweight as it inherits almost everything from the DC Simulation.

File: SimPEG/electromagnetics/static/spontaneous_potential/_utils.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import numpy as np
+import discretize
+from discretize.utils import active_from_xyz
+from scipy.interpolate import NearestNDInterpolator, LinearNDInterpolator
+
+
+def readSeepageModel(fname, mesh=None, xsurf=None, ysurf=None):
```
Comment (jcapriot) [inline SimPEG/electromagnetics/static/spontaneous_potential/_utils.py:7]: I'm not sure this has a place in SimPEG right now...

Comment (sgkang) [reply SimPEG/electromagnetics/static/spontaneous_potential/_utils.py:7]: no need to bring this in.

**Review by sgkang — APPROVED** [review]
Looking great. Your SP implementation is concise and elegant. 
Played a little bit with the sparse inversion, and it seems working well. 

It seemed the sensitivity (or Ainv) is not updated while inversion, but it would be nice if you could check. 

It would be great to add some tutorials at least for each of three source types, and further replicate Craig's field example. 

image

Above example is accessible through: https://curvenote.com/@sgkang09/researches/sp-3d-inv-charge.2

File: SimPEG/electromagnetics/static/spontaneous_potential/simulation.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    def getRHS(self):
+        return self.Vol @ self.q
+
+    def getRHSDeriv(self, source, v, adjoint=False):
+        if adjoint:
+            return self.qDeriv.T @ (self.Vol @ v)
+        return self.Vol @ (self.qDeriv @ v)
+
+    @property
+    def deleteTheseOnModelUpdate(self):
```
[the remark this answers is not in the document]
Comment (jcapriot) [reply SimPEG/electromagnetics/static/spontaneous_potential/simulation.py:81]: you got it, I also just added something that will only clear a stored sensitivity matrix if the mapping was non-linear.

File: SimPEG/electromagnetics/static/spontaneous_potential/simulation.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        # When enabling resistivity derivatives, uncomment these lines
+        # if self.rhoMap is not None:
+        #     return super().deleteTheseOnModelUpdate
+        # Need some way to determine if the `qMap` was linear...
+        # if it was non-linear, we would need to re-compute a
+        # stored jacobian matrix.
+        return []
+
+
+class CurrentDensityMap(maps.LinearMap):
```
[the remark this answers is not in the document]
Comment (jcapriot) [reply SimPEG/electromagnetics/static/spontaneous_potential/simulation.py:90]: Updated this to behave like the old code with an additional `active_cells` argument for the `CurrentDensityMap`.

**Comment by jcapriot** [comment]
Thanks Seogi for the review. I've made the requested updates and will pull this in. Feels good to not have the leftover pieces from before the simulation switchover in the codebase anymore!

We can add tutorials and such in another PR.

**Pull request merged**